### PR TITLE
feat: [CDS-71629]: Syncing Custom stage data to timescale

### PR DESCRIPTION
**Pull request by RishabhGupta34**
## Describe your changes
Changes:
- Syncing Custom stage data to timescale
- Created new timescale table

Tested locally if the data is getting synced in timescale

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50063)

### CODE REVIEW

**Comment by RishabhGupta34** [comment]
trigger uts

**Comment by RishabhGupta34** [comment]
trigger uts

File: pipeline-service/service/src/main/resources/timescale/create_pipeline_stage_table.sql [hunk 1]
```diff
@@ -0,0 +1,13 @@
+-- Copyright 2023 Harness Inc. All rights reserved.
+-- Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
+-- that can be found in the licenses directory at the root of this repository, also available at
+-- https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.
+
+BEGIN;
+
+CREATE TABLE IF NOT EXISTS public.pipeline_stage_execution (
+                                                      id text  NOT NULL,
```
Comment (anilchowdhury-harness) [inline pipeline-service/service/src/main/resources/timescale/create_pipeline_stage_table.sql:9]: lets rename it to custom_stage_execution

File: 110-change-data-capture/src/main/java/io/harness/changehandlers/PipelineStageExecutionHandler.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      long startTs = Long.parseLong(dbObject.get(StageExecutionEntityKeys.startts).toString());
+      long endTs = Long.parseLong(dbObject.get(StageExecutionEntityKeys.endts).toString());
+      long duration = endTs - startTs;
+      columnValueMapping.put("duration", Long.toString(duration));
+    }
+
+    if (dbObject.get(StageExecutionEntityKeys.stageType) != null) {
+      if ("CUSTOM_STAGE".equals(dbObject.get(StageExecutionEntityKeys.stageType).toString())) {
+        columnValueMapping.put("type", "CUSTOM");
+      } else {
```
Comment (anilchowdhury-harness) [inline 110-change-data-capture/src/main/java/io/harness/changehandlers/PipelineStageExecutionHandler.java:57]: let rename it to "Custom"

File: 110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    }
+    Map<String, String> columnValueMapping = new HashMap<>();
+    DBObject dbObject = changeEvent.getFullDocument();
+
+    if (dbObject == null) {
+      return null;
+    }
+    if (dbObject.get(StageExecutionEntityKeys.failureInfo) == null) {
+      return null;
+    }
```
Comment (anilchowdhury-harness) [inline 110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java:45]: move this to new method, validateEventData()

File: 110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    columnValueMapping.put("id", dbObject.get(StageExecutionEntityKeys.stageExecutionId).toString());
+    BasicDBObject failureInfo = (BasicDBObject) dbObject.get(StageExecutionEntityKeys.failureInfo);
+    if (failureInfo.get("errorMessage") != null) {
+      changeHandlerHelper.addKeyValuePairToMapFromDBObject(
+          failureInfo, columnValueMapping, "errorMessage", "failure_message");
+    } else if (failureInfo.get("failureData") != null) {
+      BasicDBList failureData = (BasicDBList) failureInfo.get("failureData");
+      if (!isEmpty(failureData)) {
+        BasicDBObject[] failureDataArray = failureData.toArray(new BasicDBObject[failureData.size()]);
+        if (failureDataArray[0].get("message") != null) {
```
Comment (anilchowdhury-harness) [inline 110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java:56]: check if we need to convert the list to array

Comment (RishabhGupta34) [reply 110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java]: done

File: 110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java (hunk 3 above)
Comment (RishabhGupta34) [reply 110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java]: done

File: 110-change-data-capture/src/main/java/io/harness/changehandlers/PipelineStageExecutionHandler.java (hunk 2 above)
Comment (RishabhGupta34) [reply 110-change-data-capture/src/main/java/io/harness/changehandlers/PipelineStageExecutionHandler.java:47]: done

File: pipeline-service/service/src/main/resources/timescale/create_pipeline_stage_table.sql (hunk 1 above)
Comment (RishabhGupta34) [reply pipeline-service/service/src/main/resources/timescale/create_pipeline_stage_table.sql]: done

**Comment by RishabhGupta34** [comment]
trigger uts

**Review by anilchowdhury-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50063#-:-NZ_uHwQ9911F-KiWOz9:bnfp4nl)* status: 25 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by archit-harness — APPROVED** [review]
> Reviewed 5 of 21 files at r1, 1 of 4 files at r2, 10 of 10 files at r3, 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50063)* status: 25 file(s) to review. 1. unresolved discussions.

**Comment by RishabhGupta34** [comment]
trigger uts

**Review by anilchowdhury-harness — APPROVED** [review]

**Review by archit-harness — APPROVED** [review]

**Comment by RishabhGupta34** [comment]
trigger review

**Pull request merged**